File: raspberry/SmartHome/thirdpartydevices/src/main/java/raspberry/smarthome/thirdpartydevices/xiaomi/gateway/device/THSensor.java
package raspberry.smarthome.thirdpartydevices.xiaomi.gateway.device;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Optional;

import raspberry.smarthome.thirdpartydevices.R;

public class THSensor extends ExtendedDevice {

    private OnTemperatureChangeListener onTemperatureChangeListener;
    private OnHumidityChangeListener onHumidityChangeListener;

    private Float temperature = null;
    private Float humidity = null;

    public THSensor(String sid) {
        super(sid, TEMPERATURE_HUMIDITY_SENSOR_TYPE);
    }

    public THSensor(String sid, String type) {
        super(sid, type);
    }

    public THSensor(String sid, OnTemperatureChangeListener onTemperatureChangeListener) {
        this(sid);
        setOnTemperatureChangeListener(onTemperatureChangeListener);
    }

    public THSensor(String sid, OnHumidityChangeListener onHumidityChangeListener) {
        this(sid);
        setOnHumidityChangeListener(onHumidityChangeListener);
    }

    public THSensor(String sid, OnTemperatureChangeListener onTemperatureChangeListener, OnHumidityChangeListener onHumidityChangeListener) {
        this(sid);
        setOnTemperatureChangeListener(onTemperatureChangeListener);
        setOnHumidityChangeListener(onHumidityChangeListener);
    }

    public void setOnTemperatureChangeListener(OnTemperatureChangeListener onTemperatureChangeListener) {
        this.onTemperatureChangeListener = onTemperatureChangeListener;
    }

    public void setOnHumidityChangeListener(OnHumidityChangeListener onHumidityChangeListener) {
        this.onHumidityChangeListener = onHumidityChangeListener;
    }

    @Override
    public void parseData(String cmd) {
        try {
            JSONObject o = new JSONObject(cmd);

            if(!o.isNull(STATUS_TEMPERATURE)) {
                float currTemp = Float.parseFloat(o.getString(STATUS_TEMPERATURE)) / 100;

                if(temperature != null && Math.abs(currTemp - temperature) > 0.01) {
                    Optional.ofNullable(onTemperatureChangeListener).ifPresent(onTemperatureChangeListener -> onTemperatureChangeListener.onTemperatureChanged(currTemp));
                }

                temperature = currTemp;
            }

            if(!o.isNull(STATUS_HUMIDITY)) {
                float currHumidity = Float.parseFloat(o.getString(STATUS_HUMIDITY)) / 100;

                if(humidity != null && Math.abs(currHumidity - humidity) > 0.01) {
                    Optional.ofNullable(onHumidityChangeListener).ifPresent(onHumidityChangeListener -> onHumidityChangeListener.onHumidityChanged(currHumidity));
                }

                humidity = currHumidity;
            }

            if(!o.isNull(VOLTAGE_KEY)) {
                String vol = o.getString(VOLTAGE_KEY);
                setVoltage(Float.parseFloat(vol) / 1000);
            }

        } catch (JSONException | NumberFormatException e) {
            e.printStackTrace();
        }
    }

    @Override
    public String toString() {
        return super.toString() + "\ntemperature: " + temperature + "C, humidity: " + humidity + "%";
    }

    @Override
    public Bitmap getDevicePicture(Resources resources) {
        return BitmapFactory.decodeResource(resources, R.drawable.xiaomi_th_sensor);
    }

    public interface OnTemperatureChangeListener {

        void onTemperatureChanged(float t);
    }

    public interface OnHumidityChangeListener {

        void onHumidityChanged(float h);
    }
}

File: raspberry/SmartHome/thirdpartydevices/src/main/java/smarthome/raspberry/thirdpartydevices/xiaomi/gateway/device/THSensor.kt
package smarthome.raspberry.thirdpartydevices.xiaomi.gateway.device

import org.json.JSONException
import org.json.JSONObject
import smarthome.library.common.constants.DeviceTypes.TEMPERATURE_HUMIDITY_SENSOR_TYPE
import smarthome.library.common.constants.GATEWAY_HUMIDITY_CONTROLLER
import smarthome.library.common.constants.GATEWAY_TEMPERATURE_CONTROLLER
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.constants.STATUS_HUMIDITY
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.constants.STATUS_TEMPERATURE
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.HumidityController
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.TemperatureController
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.VoltageController

open class THSensor (sid: String,
                     type: String = TEMPERATURE_HUMIDITY_SENSOR_TYPE,
                     gatewaySid: String)
    : GatewayDevice(sid, type, gatewaySid) {

    init {
        addControllers(TemperatureController(this),
                HumidityController(this),
                VoltageController(this))
    }

    override fun parseData(json: String) {
        try {
            val o = JSONObject(json)

            if (!o.isNull(STATUS_TEMPERATURE))
                getControllerByType(GATEWAY_TEMPERATURE_CONTROLLER).state = (o.getString(STATUS_TEMPERATURE).toFloat() / 100).toString() + "C"

            if (!o.isNull(STATUS_HUMIDITY))
                getControllerByType(GATEWAY_HUMIDITY_CONTROLLER).state = (o.getString(STATUS_HUMIDITY).toFloat() / 100).toString() + "%"

            setVoltage(o)

        } catch (e: JSONException) {
            reportDataParseError(e)
        }
    }

    override fun toString(): String {
        return super.toString() + "\ntemperature: " + getControllerByType(GATEWAY_TEMPERATURE_CONTROLLER).state +
                ", humidity: " + getControllerByType(GATEWAY_HUMIDITY_CONTROLLER).state
    }
}